test/Test/Data/Mutable/Array.hs

```diff
@@ -1,16 +1,112 @@
+{-# LANGUAGE LinearTypes #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE TemplateHaskell #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
 
-module Test.Data.Mutable.Array (mutArrTests) where
+-- |
+-- Tests for mutable arrays.
+--
+-- TODO:
+--  * Test failures for out of bound access
+module Test.Data.Mutable.Array
+  ( mutArrTests,
+  )
+where
 
+import qualified Data.Array.Mutable.Linear as Arr
+import Data.Unrestricted.Linear
 import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Prelude.Linear as Lin
+import Test.Tasty
+import Test.Tasty.Hedgehog (testProperty)
+import Test.Util
 
-mutArrTests :: IO Bool
-mutArrTests = checkParallel group
+-- # Exported Tests
+--------------------------------------------------------------------------------
 
-group :: Group
-group = Group "Mutable array tests" $
-  []
+mutArrTests :: TestTree
+mutArrTests = testGroup "Mutable array tests" group
 
+group :: [TestTree]
+group =
+  [ testProperty "[read i (write arr i x) = x]" testWrite,
+    testProperty "[alloc creates constant arrays]" testAlloc,
+    testProperty "[resize works correctly]" testResize
+  ]
 
+-- # Internal Library
+--------------------------------------------------------------------------------
 
+type ArrayTester a = Arr.Array a #-> Unrestricted Bool
+
+-- Given pairs of (index,val), write to an array
+makeArray :: [(Int, Int)] -> Arr.Array Int #-> Arr.Array Int
+makeArray [] arr = arr
+makeArray ((ix, val) : xs) arr = makeArray xs Lin.$ Arr.write arr ix val
+
+-- | Given an array size, create random values to insert
+randomArrVals :: Int -> Gen [(Int, Int)]
+randomArrVals size = do
+  numInsert <- Gen.int $ Range.constant 150 500
+  let numToInsert = Range.singleton numInsert
+  let genVal = Gen.int $ Range.constant 1 300
+  let genIx = Gen.int $ Range.constant 0 (size -1)
+  ixs <- Gen.list numToInsert genIx
+  vals <- Gen.list numToInsert genVal
+  return $ zip ixs vals
+
+-- # Tests
+--------------------------------------------------------------------------------
+
+-- The read of a cell we wrote to should be what we wrote.
+testWrite :: Property
+testWrite = property $ do
+  arrSize <- forAll $ Gen.int (Range.constant 1 100)
+  ix <- forAll $ Gen.int $ Range.constant 0 (arrSize - 1)
+  val <- forAll $ Gen.int $ Range.constant 0 100
+  randInserts <- forAll $ randomArrVals arrSize
+  let test = testWriteLin ix val #. makeArray randInserts
```

Comment:
Give yourself a `fromList` function with the same sort of signature as `alloc`. It will make your life simpler. Probably in the `Array` module.